File: click/python/click_payment/click_complete.py
"""Click COMPLETE so'rovi.

Click bu so'rovni "pul yechildi, endi mahsulotni ber" deb yuboradi
(yoki "foydalanuvchi bekor qildi" deb — `error` maydoniga qarab).

Biz tekshiramiz:
    - so'rov to'liqmi                       -> yo'q bo'lsa -8
    - imzo haqiqiymi                        -> yo'q bo'lsa -1
    - bunday buyurtma bormi                 -> yo'q bo'lsa -5
    - merchant_prepare_id prepare'dagi id'ga mos keladimi -> yo'q bo'lsa -6
    - summa bazadagiga mos keladimi         -> yo'q bo'lsa -2

Hammasi joyida bo'lsa buyurtmani "to'landi" deb belgilaymiz va
click_orders.on_paid() ni chaqiramiz — mahsulot o'sha yerda beriladi.

Bu faylga tegishingiz shart emas — bazangizga ulanish click_orders.py da.
"""

from __future__ import annotations

import logging
from typing import Any, Mapping

from . import click_errors as err
from . import click_orders
from .click_config import get_config
from .click_signature import ACTION_COMPLETE, check_sign
from .click_utils import amounts_match, as_int, missing_fields

logger = logging.getLogger("click")

# Complete'da prepare'dagilarga qo'shimcha merchant_prepare_id ham keladi.
REQUIRED_FIELDS = (
    "click_trans_id",
    "service_id",
    "merchant_trans_id",
    "merchant_prepare_id",
    "amount",
    "sign_time",
    "sign_string",
)


def handle_complete(data: Mapping[str, Any]) -> dict[str, Any]:
    """Click complete so'rovini qayta ishlaydi va javob dict qaytaradi."""
    missing = missing_fields(data, REQUIRED_FIELDS)
    if missing:
        logger.warning("Click complete: maydon yetishmayapti: %s", missing)
        return _response(data, 0, err.BAD_REQUEST)

    if not check_sign(data, ACTION_COMPLETE, get_config()):
        logger.warning(
            "Click complete: imzo xato (merchant_trans_id=%s)",
            data.get("merchant_trans_id"),
        )
        return _response(data, 0, err.SIGN_CHECK_FAILED)

    merchant_trans_id = str(data["merchant_trans_id"])
    click_trans_id = str(data["click_trans_id"])
    order = click_orders.find_order(merchant_trans_id)

    if order is None:
        logger.warning("Click complete: buyurtma topilmadi (%s)", merchant_trans_id)
        return _response(data, 0, err.USER_NOT_FOUND)

    # prepare'da qaytargan id bilan bir xil bo'lishi shart.
    if as_int(data.get("merchant_prepare_id")) != order.id:
        logger.warning(
            "Click complete: merchant_prepare_id mos emas (kutilgan=%s, kelgan=%s)",
            order.id,
            data.get("merchant_prepare_id"),
        )
        return _response(data, order.id, err.TRANSACTION_NOT_FOUND)

    if not amounts_match(data.get("amount"), order.amount):
        logger.warning(
            "Click complete: summa mos emas (bazada=%s, so'rovda=%s)",
            order.amount,
            data.get("amount"),
        )
        return _response(data, order.id, err.INCORRECT_AMOUNT)

    # Takroriy callback: Click javobni ololmay qayta urgan. Pul allaqachon
    # hisobga olingan — muvaffaqiyat deb javob beramiz, on_paid QAYTA
    # chaqirilmaydi.
    if order.status == click_orders.STATUS_PAID:
        logger.info("Click complete: takroriy callback (%s)", merchant_trans_id)
        return _response(data, order.id, err.SUCCESS)

    if order.status == click_orders.STATUS_CANCELLED:
        return _response(data, order.id, err.TRANSACTION_CANCELLED)

    # Click o'zi bekor qilish/xato haqida xabar bergan.
    click_error = as_int(data.get("error"))
    if click_error != 0:
        click_orders.mark_cancelled(order, click_trans_id)
        order.status = click_orders.STATUS_CANCELLED
        logger.info(
            "Click complete: to'lov bekor qilindi (%s, click error=%s)",
            merchant_trans_id,
            click_error,
        )
        _safe_call(click_orders.on_cancelled, order, "on_cancelled")
        return _response(data, order.id, err.TRANSACTION_CANCELLED)

    # Asosiy holat: to'lov muvaffaqiyatli.
    #
    # mark_paid() faqat HAQIQATAN pending -> paid o'tkazgan bo'lsa True
    # qaytaradi. Parallel kelgan ikkinchi callback False oladi va mahsulot
    # ikki marta berilmaydi.
    if not click_orders.mark_paid(order, click_trans_id):
        logger.info(
            "Click complete: boshqa callback ulgurdi (%s) — on_paid o'tkazib yuborildi",
            merchant_trans_id,
        )
        return _response(data, order.id, err.SUCCESS)

    order.status = click_orders.STATUS_PAID
    logger.info(
        "Click complete OK (buyurtma=%s, id=%s, summa=%s)",
        merchant_trans_id,
        order.id,
        order.amount,
    )
    _safe_call(click_orders.on_paid, order, "on_paid")

    return _response(data, order.id, err.SUCCESS)


def _safe_call(func, order, name: str) -> None:
    """Hodisa funksiyasini chaqiradi; xato bo'lsa loglaydi, javobni buzmaydi.

    Nega xatoni yutamiz? Bu nuqtaga yetganda pul yechilgan va buyurtma
    "paid" deb belgilangan. Click'ga xato qaytarsak u qayta uradi — lekin
    yuqorida "allaqachon to'langan" bo'lib SUCCESS oladi, ya'ni on_paid
    baribir qayta ishlamaydi. Shuning uchun to'g'ri yo'l: xatoni loglab,
    keyin qo'lda hal qilish.
    """
    try:
        func(order)
    except Exception:
        logger.exception(
            "Click %s() xatosi (buyurtma=%s) — to'lov 'paid' holicha qoldi, "
            "qo'lda tekshiring",
            name,
            order.merchant_trans_id,
        )


def _response(
    data: Mapping[str, Any], merchant_confirm_id: int, error: int
) -> dict[str, Any]:
    """Click kutadigan javob shakli.

    Diqqat: prepare'da `merchant_prepare_id`, complete'da esa
    `merchant_confirm_id` deb nomlanadi.
    """
    return {
        "click_trans_id": as_int(data.get("click_trans_id")),
        "merchant_trans_id": str(data.get("merchant_trans_id", "")),
        "merchant_confirm_id": int(merchant_confirm_id),
        "error": int(error),
        "error_note": err.error_note(error),
    }

File: click/typescript/click-payment/click-complete.ts
/**
 * Click COMPLETE so'rovi.
 *
 * Click bu so'rovni "pul yechildi, endi mahsulotni ber" deb yuboradi
 * (yoki "foydalanuvchi bekor qildi" deb — `error` maydoniga qarab).
 *
 * Biz tekshiramiz:
 *     - so'rov to'liqmi                       -> yo'q bo'lsa -8
 *     - imzo haqiqiymi                        -> yo'q bo'lsa -1
 *     - bunday buyurtma bormi                 -> yo'q bo'lsa -5
 *     - merchant_prepare_id prepare'dagi id'ga mos keladimi -> yo'q bo'lsa -6
 *     - summa bazadagiga mos keladimi         -> yo'q bo'lsa -2
 *
 * Hammasi joyida bo'lsa buyurtmani "to'landi" deb belgilaymiz va
 * onPaid() ni chaqiramiz — mahsulot o'sha yerda beriladi.
 *
 * Bu faylga tegishingiz shart emas — bazangizga ulanish click-orders.ts da.
 */

import { getConfig } from "./click-config.js";
import { ClickError, errorNote } from "./click-errors.js";
import {
  orders as defaultOrders,
  type ClickOrder,
  type ClickOrdersAdapter,
} from "./click-orders.js";
import { ACTION_COMPLETE, checkSign } from "./click-signature.js";
import {
  amountsMatch,
  asInt,
  clickLog,
  missingFields,
  type ClickRequestData,
} from "./click-utils.js";

/** Complete'da prepare'dagilarga qo'shimcha merchant_prepare_id ham keladi. */
export const COMPLETE_REQUIRED_FIELDS = [
  "click_trans_id",
  "service_id",
  "merchant_trans_id",
  "merchant_prepare_id",
  "amount",
  "sign_time",
  "sign_string",
] as const;

export interface ClickCompleteResponse {
  click_trans_id: number;
  merchant_trans_id: string;
  merchant_confirm_id: number;
  error: number;
  error_note: string;
}

/**
 * Click complete so'rovini qayta ishlaydi va javob obyektini qaytaradi.
 *
 * @param data   So'rovdagi maydonlar (form yoki JSON — farqi yo'q)
 * @param orders Bazangizga ulanish (odatda berilmaydi).
 */
export async function handleComplete(
  data: ClickRequestData,
  orders: ClickOrdersAdapter = defaultOrders,
): Promise<ClickCompleteResponse> {
  const missing = missingFields(data, COMPLETE_REQUIRED_FIELDS);
  if (missing.length > 0) {
    clickLog("warn", "complete: maydon yetishmayapti", { missing });
    return response(data, 0, ClickError.BAD_REQUEST);
  }

  if (!checkSign(data, ACTION_COMPLETE, getConfig())) {
    clickLog("warn", "complete: imzo xato", {
      merchant_trans_id: data.merchant_trans_id,
    });
    return response(data, 0, ClickError.SIGN_CHECK_FAILED);
  }

  const merchantTransId = String(data.merchant_trans_id);
  const clickTransId = String(data.click_trans_id);
  const order = await orders.findOrder(merchantTransId);

  if (order === null) {
    clickLog("warn", "complete: buyurtma topilmadi", { merchantTransId });
    return response(data, 0, ClickError.USER_NOT_FOUND);
  }

  // prepare'da qaytargan id bilan bir xil bo'lishi shart.
  if (asInt(data.merchant_prepare_id) !== order.id) {
    clickLog("warn", "complete: merchant_prepare_id mos emas", {
      kutilgan: order.id,
      kelgan: data.merchant_prepare_id,
    });
    return response(data, order.id, ClickError.TRANSACTION_NOT_FOUND);
  }

  if (!amountsMatch(data.amount, order.amount)) {
    clickLog("warn", "complete: summa mos emas", {
      bazada: order.amount,
      sorovda: data.amount,
    });
    return response(data, order.id, ClickError.INCORRECT_AMOUNT);
  }

  // Takroriy callback: Click javobni ololmay qayta urgan. Pul allaqachon
  // hisobga olingan — muvaffaqiyat deb javob beramiz, onPaid QAYTA
  // chaqirilmaydi.
  if (order.status === "paid") {
    clickLog("info", "complete: takroriy callback", { merchantTransId });
    return response(data, order.id, ClickError.SUCCESS);
  }

  if (order.status === "cancelled") {
    return response(data, order.id, ClickError.TRANSACTION_CANCELLED);
  }

  // Click o'zi bekor qilish/xato haqida xabar bergan.
  const clickError = asInt(data.error ?? 0);
  if (clickError !== 0) {
    await orders.markCancelled(order, clickTransId);
    order.status = "cancelled";

    clickLog("info", "complete: to'lov bekor qilindi", {
      merchantTransId,
      clickError,
    });

    await safeCall(() => orders.onCancelled(order), "onCancelled", order);

    return response(data, order.id, ClickError.TRANSACTION_CANCELLED);
  }

  // Asosiy holat: to'lov muvaffaqiyatli.
  //
  // markPaid() faqat HAQIQATAN pending -> paid o'tkazgan bo'lsa true
  // qaytaradi. Parallel kelgan ikkinchi callback false oladi va mahsulot
  // ikki marta berilmaydi.
  const won = await orders.markPaid(order, clickTransId);
  if (!won) {
    clickLog("info", "complete: boshqa callback ulgurdi — onPaid o'tkazib yuborildi", {
      merchantTransId,
    });
    return response(data, order.id, ClickError.SUCCESS);
  }

  order.status = "paid";

  clickLog("info", "complete OK", {
    merchantTransId,
    orderId: order.id,
    amount: order.amount,
  });

  await safeCall(() => orders.onPaid(order), "onPaid", order);

  return response(data, order.id, ClickError.SUCCESS);
}

/**
 * Hodisa funksiyasini chaqiradi; xato bo'lsa loglaydi, javobni buzmaydi.
 *
 * Nega xatoni yutamiz? Bu nuqtaga yetganda pul yechilgan va buyurtma
 * "paid" deb belgilangan. Click'ga xato qaytarsak u qayta uradi — lekin
 * yuqorida "allaqachon to'langan" bo'lib SUCCESS oladi, ya'ni onPaid
 * baribir qayta ishlamaydi. Shuning uchun to'g'ri yo'l: xatoni loglab,
 * keyin qo'lda hal qilish.
 */
async function safeCall(
  fn: () => Promise<void>,
  name: string,
  order: ClickOrder,
): Promise<void> {
  try {
    await fn();
  } catch (e) {
    clickLog(
      "error",
      `${name}() xatosi — to'lov "paid" holicha qoldi, qo'lda tekshiring`,
      {
        merchantTransId: order.merchantTransId,
        xato: e instanceof Error ? e.message : String(e),
      },
    );
  }
}

/**
 * Click kutadigan javob shakli.
 *
 * Diqqat: prepare'da `merchant_prepare_id`, complete'da esa
 * `merchant_confirm_id` deb nomlanadi.
 */
function response(
  data: ClickRequestData,
  merchantConfirmId: number,
  error: number,
): ClickCompleteResponse {
  return {
    click_trans_id: asInt(data.click_trans_id),
    merchant_trans_id: String(data.merchant_trans_id ?? ""),
    merchant_confirm_id: merchantConfirmId,
    error,
    error_note: errorNote(error),
  };
}
